### DISCUSSION TITLE
Stream word-based field unpacker

### DISCUSSION THREAD

**Discussion by dokleina** [SpinalHDL/SpinalHDL#787] [Ideas]
I've had this idea for a while now and I wanted to get some comments and thoughts on it.

I'd like to be able to describe the fields of a bundle in terms of words and bit positions within those words. The description could be used by a component to unpack the fields from a stream as they flow into/through it.

A mock up of how this might look in code:
```
val My_Data_Structure = new UnpackableBundle {
    val field_A = Bits(16 bits) // 16 bit field
    val flag_B = Bool offset(3) // Flag at bit 1
    val flag_C = Bits(8 bits) offset(8) // 8 bit field, starting at bit 8
}
```

I'm unsure of where the logic for unpacking would be generated. My first idea is to want some kind of Factory-pattern that lives with the extended/unpackable bundle. which might look like:
```
val myUnpackedData = My_Data_Structure.unpack(someInputStream)
```

To represent unpacked fields, the items within the returned bundle would have `valid` (or something similar):
```
when(myUnpackedData.field_A.valid) {
     // Do something with field_A
}
```

There are a few other things that need to be thought through:

- Should word width be a part of the structure description? (Would `new UnpackableBundle(<number> bits)` be user-friendly?)
- If a field is too long, should it be wrapped or raise an error at build time?
- Would endianness need to be considered?
- How should this type (or bundle mix-in?) be implemented?

It's likely I've missed a few points, but I'll try to answer questions any additional questions people might have.

Thank you for your time.

**Comment by Dolu1990** [comment]
Hi ^^

Does UnpackableBundle represent something kind of a mix betwen C struct and C union ?
I mean, from unpack/pack standpoint things can overlap, but from a instance standpoint, each field has its dedicated wires. ?

> Would endianness need to be considered?

Yes ^^

**Reply by dokleina** [reply]
I suppose it could be thought of as kind of like a C struct and union. The output would be a Bundle with all of the fields declared in the unpackable bundle form, so they would have their own wires. The unpackable part only serves to describe where the field bits would be sourced from within words and across words of an input stream.

Overlapping would only need to be disallowed/an error if the descriptor is also used to pack a Bundle into a stream. In that case, overlaps would lead to double-driven bits.

It might make sense to make a reversible descriptor which disallows overlaps but packs and unpacks, and a solely unpacking descriptor that doesn't care about overlaps.

**Reply by Dolu1990** [reply]
One question would be : 
"Should the location information be stored into the stream ? Or should it just be a external LinkedHashMap[Data, Int]" ?

I mean, if it is realy related to serialiazing/unserializing from stream, i would say this should maybe not be too tightly coupled ?
We can imagine having LinkedHashMap[Data, Int], or a class in which you can push Data with given offset to constructe a mapping ?

**Reply by dokleina** [reply]
I definitely see this as class to be consumed by a Component. There will be an included de/serializer component to handle generic Stream -> Bundle and vice versa. Components would be free to accept an un/packable-bundle as part of their parameters.

I'm thinking that the un/packable-bundle itself will be convertible to/from a standard Bundle with the same Data names. I haven't had time to look into this too much, but I believe I could create a Bundle from a Hardtype and the Data names. That might look like:
```
val My_Data_Structure = new UnpackableBundle {
    ... // Data stuff here
}
val myOutputData = My_Data_Structure.asBundle // Make a new object with the underlying Bundle from My_Data_Structure
```

**Reply by Dolu1990** [reply]
Hmm, trying to think about the low level of it, without API magic (for now) : 

```scala
val stream = Stream(Bits(8 bits))
val rgb = Rgb(5,6,5)
val unpacker = StreamUnpacker[Rgb](
  hardType = Rgb(5,6,5),
  //Layout is a landa function to return the mapping between a payload elements and the location in the stream
  layout = payload => List(   
    payload.r -> 0, //first byte
    payload.g -> 8, //seconde byte
    payload.b -> 16 //third byte
  )
)

unpacker.io.input << stream
rgb := unpacker.io.output
```

Does it make sense in some ways ?

**Reply by dokleina** [reply]
Yeah, that makes sense.

I'm not sure how I feel about making a spec for an existing Bundle. It might make it too easy to have the Bundle and the Bundle's stream unpacking spec diverge.

_Edit: I see now that it's a function called "StreamUnpacker" and not a class. I think that's exactly what this needs if layout can be extracted from the unpack describer bundle._

**Reply by Dolu1990** [reply]
hmm, i'm not sure we should realy assume that a stream need to be deserialized into a bundle, but instead into a spec which map Data -> Int

So something more flexible would be (from a low level API) : 

```scala
val stream = Stream(Bits(8 bits))
val rgb = Rgb(5,6,5)
val alpha = UInt(16 bits)
val unpacker = StreamUnpacker[Rgb](
  layout = List(   
    rgb.r -> 0, //first byte
    rgb.g -> 8, //seconde byte
    rgb.b -> 16 //third byte
    alpha(7 downto 0) -> 24 //forth byte
    alpha(15 downto 8) -> 32 //fift byte
  )
)
```

Do you see what i mean ? This way you don't have to force yourself to use a bundle with everything, you can compose stuff.

**Reply by dokleina** [reply]
True, this doesn't need to be limited to unpacking into a Bundle. I like that example.

I'm focused on a Bundle form so that I can describe the fields of the Stream to be deserialized and associate a name with them. I figure extending Bundle gives me the naming association that I want.

An example of what I wanted to do with the Bundles:
```scala
class DMADescriptor extends UnpackableBundle(8 bits) {
    val srcAddr = UInt(32 bits)
    val destAddr = UInt(32 bits)
    val nextDesc = UInt(32 bits)
    // Etc.
}

// Memory Bus input from somewhere
val memBus = Stream(Bits(8 bits))

// Unpacker component
val memUnpacker = StreamUnpacker[DMADescriptor]()
memUnpacker.io.inBus << memBus
val readDescriptor = memUnpacker.io.out

when(memUnpacker.done) {
    // Do something when the whole DMADescriptor has been unpacked
}

// Extra Goal: Add a `.fire` to each field as it gets pulled from the stream
when(readDescriptor.srcAddr.fire) {
    // Do something when just the `srcAddr` has been fully unpacked
}
```

**Reply by Dolu1990** [reply]
hmm then i would kind of realy tends more into the "layout = List(   xxx -> yyy, ...) way.

Where memUnpacker would provide the "done" and a fire for each element of the layout via a LinkedHashMap.

**Reply by dokleina** [reply]
Alright. I'll take a go at implementing the List layout and see how it turns out.

**Reply by dokleina** [reply]
Update: I made some progress on this in #814. It's using the List of Data -> Bit and it provides the "done" flags for each field, as well as an "all done" flag.

I'd still really like a way to create a single source of layout specification. I'm still going through ideas of how that would look and work.
